### ISSUE TITLE
Roadmap during GSoC

### ISSUE THREAD

**Issue by nd-02110114** [deepchem/jaxchem#1]
I wrote down the roadmap of this project in GSoC.

# Roadmap
- The 1st evaluation
- The 2nd evaluation
- The final evaluation
- After the GSoC

## The 1st evaluation (June 01 ~ July 05)
In this period, I will focus on making the GNN (GCN) benchmark. 
~Currently, I choose the Gated Graph Neural Network as a benchmark model, but I may change the model to a more simple GCN model like [this blog tutorial](http://gcucurull.github.io/deep-learning/2020/04/20/jax-graph-neural-networks/).~

I chose the GCN model !

~- [ ] Implement the sparse pattern Gated Graph Neural Network~
~- [ ] Implement the pad pattern Gated Graph Neural Network~
- [x] Implement the pad pattern GCN
- [x] Make the google colab tutorials about the models described above

GCN implementation : https://github.com/gcucurull/jax-gcn

We could check the difference between the sparse pattern and the pad patten in the following links. I want to try compare the two patterns and DGL models about inference speed or memory efficiency.

- https://github.com/john-bradshaw/GNN
- https://www.slideshare.net/pfi/20190930-pfn-internship-2019-extension-of-chainerchemistry-for-large-amp-sparse-graphkenshin-abe

## The 2nd evaluation (July 06 ~ Aug 02)

~In this period, I will focus on implementing the Crystal Graph Convolutional Neural Network (CGCNN) and supporting material data and modelings. However, I  may try to implement the more physical model like [Lattice Convolutional Neural Network](https://github.com/VlachosGroup/lcnn) related to surface reaction instead of the CGCNN.~

~- [ ] Implement the Crystal Graph Convolutional Neural Network (CGCNN)~
~- [ ] Implement utility functions of Material Project Data in moleculenet~
~- [ ] Make the google colab tutorial about CGCNN~

According to the roadmap, I’ll be working for implementing the CGCNN model. In this period, I will focus on supporting inorganic crystal data before Please confirm the details below.

- [x] Build the JAXChem document (7/6 - 7/12)
- [ ] Implement a dataloader for crystal data and a new graph data class (7/13 - 7/20)
- [ ] Implement the featurizer for CGCNN (7/20 - 7/27)
- [ ] Implement the CGCNN model (7/27 - 8/3)

CGCNN implementation : https://github.com/txie-93/cgcnn

## The final evaluation (Aug 03 ~ Aug 31)
In this period, I will focus on implementing the attention model. I think the attention model tend to be a little complex network compared with GCN, so it is a good example of JAX Chem.

- [ ] Implement the Molecular Attention Transformer (MAT)
- [ ] Make the google colab tutorial about MAT

MAT implementation : https://github.com/ardigen/MAT

## ~After the GSoC~ -> I'm working for these things during GSoC

(Thank you for advice!)

After the GSoC, I will focus on writing documents and tests.

- [x] Write documents
- [x] Write unit tests -> I'm writing the unit test from 1st evaluation period
- [x] Set up a CI -> I created the CI configuration in 1st evaluation period
 
# Development environments

## CI configuration

I use the GitHub Action.

- pipeline
  - build check
  - unit test : **pytest**
  - type check : **mypy**
  - linter : **flake8**

**Comment by vsomnath** [comment]
Hi Daiki, 

Nice to have the roadmap here!

Just one suggestion, you should write documentation and tests in parallel with the code. This is good practice and forces you to think both about the API and the associated corner cases.

**Comment by nd-02110114** [comment]
@vsomnath Thanks you for your comment!
@lilleswing @rbharath 

I get used to JAX by implementing Tox21 example #3 and I take some advice form vsomnath.
So, I rethink the plan in June. Please check it and give me feedback on my plan.

##  Update the plan in June
- [ ] Modify update function of GCN model **(6/9~6/14)**
    - Original GCN normalizes adjacency matrix, but current my implementation doesn't normalize
    - I seem this modification is easy.
- [ ] Write test code **(6/19~6/14)**
  -  vsomnath's advice! It's better to have test codes before refactoring
- [ ] Refactor GCN model by using Haiku **(6/15~6/21)**
  - pure JAX implementation is a little hard to manage params (see `init_fun` in GCN model)
  - `init_fun` disrupts our focus on forward computation implementation.
  - Haiku provide base class for making us focus on forward computation implementation.
- [ ] Comparison training performance with deepchem or dgl **(6/21~6/25)**
   - Karl's advice! We should confirm the performance of JAX.
   - I will use Tox21 or QM9 dataset (I think QM9 is better because Tox21 is a small dataset)
- [ ] add type annotation  **(6/25~6/29)**
   - vsomnath's advice! in #3
   - I thinks type annotation is good because it makes our codes more readable and maintainable
- [ ] Implement sparse pattern GIN model **(6/30~)**
    -  GIN is better compared with GGNN because GIN shows good accuracy for many chemical tasks (see [Open Graph Benchmark result](https://ogb.stanford.edu/docs/leader_graphprop/))
    - current GCN model is a pad pattern model, so I think the sparse pattern sample model is needed.

## About Haiku
Haiku : https://github.com/deepmind/dm-haiku

- Pros 
   - User focus on forward computation implementation
   - Haiku has an object oriented design, so many python users are easy to read
- Cons
   - Haiku is beta and there are a few maintainers

**Comment by nd-02110114** [comment]
> Write test code

What kind of test should I write...? Does anyone know a good sample code?

**Comment by rbharath** [comment]
The new plan looks good! Here's a couple thoughts on tests to write, in increasing complexity

1) Shape tests: Check that each layer transforms shapes in the expected way
2) Overfit tests: A good sanity check for a model is checking that you can overfit (memorize) a small dataset. That is, make a sample dataset of 10 datapoints. Check that your model can get 100% training accuracy on this dataset.
3) Benchmark tests: These wouldn't be unit tests, but check that you can get the same accuracy on Tox21 with the new GCNs as the DeepChem or DGL GCNs

**Comment by nd-02110114** [comment]
I posted the more detail plans from now to the end of June.

- [ ] Implement sparse pattern GCN model. (**6/20~22**)
- [ ] add QM9 example (**6/24~25**)
- [ ] compare the training time among jaxchem, DGL, DeepChem and PyTorch Geometric (**6/26~28**)
- [ ] add colab notebook example (**6/30**)
- [ ] post Blog post in deepchem forum (**6/30 ~ 7/1**)
